### PR TITLE
[WIP] Profile GC

### PR DESCRIPTION
**Pull request by PalumboN**
Here is the work related to GC profiling (used for the IWST paper).

### How to use it?

1. Build the VM
1. Run any program using the generated VM (as normal)
1. A file called `scavenge.log` (but it has info about the FullGCs also) will be generated in the current directory
1. You can plot some charts from that log file using https://github.com/tesonep/benchy/pull/23

### WIP
Still need some work to be integrated.
I also want to be sure that performance is not affected (significantly) if the profiler is off.

### CODE REVIEW

File: smalltalksrc/VMMaker-Tools/VMTrace.class.st [hunk 1]
```diff
@@ -0,0 +1,58 @@
+Class {
+	#name : #VMTrace,
+	#superclass : #Object,
+	#instVars : [
+		'messageNodes'
+	],
+	#category : #'VMMaker-Tools'
+}
```
Comment (PalumboN) [inline smalltalksrc/VMMaker-Tools/VMTrace.class.st:8]: This class was used for execution trace analysis. Not related to this work.

I'll remove it (but save it in some place before)

File: smalltalksrc/VMMaker/SpurGenerationScavenger.class.st [hunk 2]
```diff
@@ -692,7 +692,7 @@ SpurGenerationScavenger >> logEndScavenge [
 SpurGenerationScavenger >> logScavenge [
 	<inline: #always>
 	scavengeLog ifNotNil:
-		[self writeScavengeLog]
+		[self writeScavengeLogAsSTON]
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/SpurGenerationScavenger.class.st:702]: Maybe we want to control this 🤔

File: smalltalksrc/VMMaker/SpurGenerationScavenger.class.st [hunk 3]
```diff
@@ -792,7 +792,9 @@ SpurGenerationScavenger >> newSpaceStart: startAddress newSpaceBytes: totalBytes
 	self initFutureSpaceStart.
 	manager initSpaceForAllocationCheck: (self addressOf: eden) limit: eden limit.
 
-	tenuringProportion := 0.9
+	tenuringProportion := 0.9.
+	
+	self openScavengeLog
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/SpurGenerationScavenger.class.st:815]: TODO: open logs from the image side

File: smalltalksrc/VMMaker/SpurGenerationScavenger.class.st [hunk 4]
```diff
@@ -1467,3 +1469,13 @@ SpurGenerationScavenger >> writeScavengeLog [
...
 	scavengeLog fflush
 ]
+
+{ #category : #logging }
+SpurGenerationScavenger >> writeScavengeLogAsSTON [
+
+	<inline: #never>
+	scavengeLogRecord deltaUsecs: manager statSGCDeltaUsecs.
+	self writeSTON: scavengeLogRecord toFile: scavengeLog.
+	self fprintf: scavengeLog _: '\n---\n'.
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/SpurGenerationScavenger.class.st:1512]: Duplicated

File: smalltalksrc/VMMaker/SpurMemoryManager.class.st [hunk 5]
```diff
@@ -12954,6 +12958,18 @@ SpurMemoryManager >> updateFullGCStats [
...
+		fullGCs: statFullGCs;
+		fullUsecs: statFullGCUsecs;
+		compactionUsecs: statCompactionUsecs;
+		sweepUsecs: gcSweepEndUsecs;
+		markUsecs: statMarkUsecs.
+	
+	self writeSTON: fullGCLogRecord toFile: scavenger logStream.
+	self fprintf: scavenger logStream _: '\n---\n'.
+	
+	scavenger logStream fflush
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/SpurMemoryManager.class.st]: `if` missing (similar to Scavenger)

**Comment by guillep** [comment]
Migrated to https://github.com/evref-inria/pharo-vm/pull/5

**Pull request closed without merging**